### PR TITLE
fuzzgen: Generate compiler flags

### PR DESCRIPTION
**Pull request by afonso360**
👋 Hey,

This PR allows fuzzgen to generate compiler flags and test using those. We only allow compiler flags that shouldn't change the behaviour of the code.

I wanted to submit this now so that we could fuzz PR's such as #4953 and #5004 that require optimizations enabled.

However, when testing this it has found an issue with one of the passes (I haven't yet narrowed it down), but I don't want to start fixing it if we are going to change pretty much everything once #4953 lands. I'm opening this as a draft, and lets re test this once #4953 lands.

<details>
<summary>The actual bug manifests as following:</summary>

Fails with:
```
target/aarch64-unknown-linux-gnu/release/cranelift-fuzzgen: Running 1 inputs 1 time(s) each.
Running: fuzz/artifacts/cranelift-fuzzgen/crash-00e37c2858c35c573ed85d618ea5d0c75ae06fda
thread '<unnamed>' panicked at 'assertion failed: `(left != right)`
  left: `v137`,
 right: `v137`: Aliasing v137 to v134 would create a loop', cranelift/codegen/src/ir/dfg.rs:322:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

And I cannot reproduce via the CLIF file since that runs into https://github.com/bytecodealliance/wasmtime/issues/4875#issuecomment-1239314171 (which I kinda forgot about 😅 ). But I'm going to try and investigate that one in the mean time.
</details>

This also enables regalloc_checker as we discussed in #4979.

cc: @jameysharp

### CODE REVIEW

**Comment by afonso360** [comment]
This PR had some performance issues:

* The regalloc checker started consuming a bunch of time (around 45%) so we now try to minimize runs with it (10% with the current iteration).
 
* I've also had to limit the amount of runs that we allow, at some point the fuzzer decided that spending 100% of the time in the interpreter was a good idea, and started generating huge inputs, that were in the order of 20k runs each. (up to 130k!)

execs/s seems way better than before, which is nice.


And for those who are interested, here's a flamegraph of where we spend time. This is about 1min at 1Khz around 190k inputs into my normal benchmark run:

flamegraph all

File: cranelift/fuzzgen/src/lib.rs
Comment (cfallin) [inline cranelift/fuzzgen/src/lib.rs:61]: Does it make sense to add `riscv64` here too (or are you worried it would find too many issues)?

Actually in the spirit of your enum `.all()` method above -- perhaps we can define a `cranelift_codegen::isa::all()` that returns all `Triple`s compiled into the build, and use that here? (Feel free to drop that idea though if it's more than ~20 lines of complexity or so.)

File: cranelift/fuzzgen/src/lib.rs
Comment (cfallin) [inline cranelift/fuzzgen/src/lib.rs:246]: It probably would make sense to leave a comment in `cranelift/codegen/meta/src/shared/settings.rs` where all of these options are defined to note "please add to fuzzing list ..." as a step when adding new otions?

File: cranelift/fuzzgen/src/lib.rs
Comment (cfallin) [inline cranelift/fuzzgen/src/lib.rs:261]: Some of these can probably be excluded, based on knowledge of what they do and likely payoff from fuzzing time:

* `machine_code_cfg_info` generates additional metadata for the embedder but this doesn't feed back into compilation anywhere; we can leave it on unconditionally to make sure the generation doesn't panic
* `enable_simd`, `enable_float`, `enable_atomics` will cause panics in compilation if an input uses one of those features and it isn't enabled, but don't change compilation otherwise; so we can probably turn them on unconditionally (we're not, e.g., trying to test that a SIMD input causes panics with the feature off)

**Review by cfallin — APPROVED** [review]
Generally looks great, thanks! Some thoughts below on specifics but I'm happy to see this landed whenever you think it's appropriate to actually start fuzzing.

File: cranelift/fuzzgen/src/lib.rs
Comment (afonso360) [reply cranelift/fuzzgen/src/lib.rs:261]: Agreed, let me know if any other flags come to mind!

> `enable_simd`, `enable_float`, `enable_atomics` will cause panics in compilation if an input uses one of those features and it isn't enabled,

Enabling them unconditionally sounds good.

But at least for `enable_floats` it hasn't panicked yet and we almost always insert floats for larger programs, so we might be silently allowing them anyway. (Haven't tested this yet)

File: cranelift/fuzzgen/src/lib.rs
Comment (afonso360) [reply cranelift/fuzzgen/src/lib.rs:62]: This ended up not being very easy to do, I might try it again later. But either way I've added `riscv64` to the target list.

**Comment by afonso360** [comment]
@cfallin What do you think about merging this with the `use_egraphs` flag commented out and enabling that on a separate PR?

It's getting a bit annoying having to rebase my other work on this PR just to fuzz it. I've fuzzed this for about an hour on both AArch64 and x86 without `use_egraphs` and it seems stable.

**Comment by cfallin** [comment]
Oh, for sure, I didn't know we were waiting on something here; sorry! I was waiting for your go-ahead to merge. FWIW the PR is still marked as a draft; go ahead and transition it to "Ready for review" when you want me to merge.

I actually think maybe it's fine to include `use_egraphs`; better to get the fuzzbugs coming sooner than later, and I'm happy to take them as they come in. But I'm open to the other approach if you think otherwise.

**Comment by afonso360** [comment]
>Oh, for sure, I didn't know we were waiting on something here; sorry! 
I wasn't really, I just got a bit annoyed today at having to do a rebase for the third time, that's where that suggestion came from!

I'm okay with merging this as is and letting fuzzgen report issues as we go along.

**Pull request merged**
